theseus/optimizer/autograd/lu_cuda_sparse_autograd.py

```diff
@@ -0,0 +1,148 @@
+        ctx.solver_context = solver_context
+        ctx.damping_alpha_beta = damping_alpha_beta
+
+        # HACK: allows to check if the context has been reused (and overwritten)
+        ctx.factor_id = solver_context.factor_id if check_factor_id else None
+
+        return x
+
+    # Let v row vector, and w column vector of dimension n, m, and
+    # A an nxm matrix. Then
+    #   v * A * w = Sum(v_i * A_ij * w_j) = [v (X) w] . A
+    # Where by [v (X) w] we mean the nxm matrix which is the
+    # tensor product of v and w, and "." is the componentwise
+    # dot product of the two nxm matrices.
+    #
+    # Now, we have
+    #      At * A * x = At * b
+    #
+    # Therefore if A, b, x are parametrized (eg. by u) we have deriving
+    #
+    # (i)  At'*A*x + At*A'*x + At*A*x' = At'*b + At*b'
+    #
+    # indicating A'=dA/du, b'=db/du, x'=dx/du
+    #
+    # Now, assume we have a function f of x, and G = df/dx be the
+    # gradient that we consider a row vector, so G*x' is df/du.
+    #
+    # To compute df/db and df/dA, make x' explicit in the (i):
+    #
+    #      x' = (At * A)^{-1} (At*b' + At'*b - At'*A*x - At*A'*x)
+    #
+    # So multiplying by the row vector G we have
+    #
+    #   G*x' = G*(At*A)^{-1}*At * b' + G*(At*A)^{-1} * (At'*b - At'*A*x - At*A'*x)
+    #        = H * At * b' + H * At' * (b - A*x) - H * At * A' * x
+    #        = (H*At) * b' + [H (X) (b-A*x)] . At' - [H*At (X) x] . A'
+    #        = (H*At) * b' + [(b-A*x) (X) H - H*At (X) x] . A'
+    # after putting H = G*(At*A)^{-1} for convenience.
+    #
+    # Therefore after switching to column vectors we have
+    #   df/db = A*H
+    # (where H = (At*A)^{-1}*G), while
+    #   df/dA = (b-A*x) (X) H - A*H (X) x
+    # The two tensor products means that to compute the gradient of
+    # a block of A we have to multiply entries taken from (b-A*x) and H,
+    # and blocks taken from A*H and x.
+    #
+    # Here we assume we are provided x and H after the linear solver
+    # has been applied to Atb and the gradient G.
+
+    # With (large) multiplicative damping, as above with extra terms:
+    #      x' = ... - (AtA_damped)^{-1} * alpha*AtA_diag'*x
+    # So multiplying by the row vector G we have
+    #      G*x' = ... - H * alpha * AtA_diag' * x
+    # Note that '...' the part multiplying H[j]*alpha*x[j] is AtA_diag'[j], ie
+    # 2 times the scalar product of A's an (A')'s j-th colum. Therefore
+    # (A')'s j-th colum is multiplying A's j-th colum by 2*H[j]*alpha*x[j]
+    @staticmethod
+    def backward(ctx, grad_output):
+
+        # HACK: check if the context has been reused (and overwritten)
+        if ctx.factor_id is not None and ctx.factor_id != ctx.solver_context.factor_id:
+            raise RuntimeError(
+                "Factoring context was overwritten! Increase the number of contexts"
+            )
+
+        batch_size = grad_output.shape[0]
+        targs = {"dtype": grad_output.dtype, "device": "cuda"}  # grad_output.device}
```

Comment:
We don't have to do it in this PR, but aside from a few lines like this one, we might be able to pull in a lot of common code between LU sparse and Cholmod sparse.

This could also work for the solver class itself by having a class hierarchy
`LinearSolver` -> `SparseSolver` ->  `CholmodSparseSolver` and `LUCudaSparseSolver`

cc @luisenp

Reply:
Talking about autograd function or the solver itself? the solver classes look to small to find any relevant common code... the autograd function I would say yes, even if the way to merge some functionality will certainly require some thoughts.

Reply:
Autograd function does have a lot of shared code, but I think it's worth considering the solver too.